src/command/backup/backup.c

```diff
@@ -0,0 +1,1906 @@
+    {
+        // Check paths
+        // -------------------------------------------------------------------------------------------------------------------------
+        case storageTypePath:
+        {
+            // If the path was not found remove it
+            if (manifestPathFindDefault(resumeData->manifest, manifestName, NULL) == NULL)
+            {
+                LOG_DETAIL_FMT("remove path '%s' from resumed backup", strPtr(storagePathP(storageRepo(), backupPath)));
+                storagePathRemoveP(storageRepoWrite(), backupPath, .recurse = true);
+            }
+            // Else recurse into the path
+            {
+                BackupResumeData resumeDataSub = *resumeData;
+                resumeDataSub.manifestParentName = manifestName;
+                resumeDataSub.backupPath = backupPath;
+
+                storageInfoListP(
+                    storageRepo(), resumeDataSub.backupPath, backupResumeCallback, &resumeDataSub, .sortOrder = sortOrderAsc);
+            }
+
+            break;
+        }
+
+        // Check files
+        // -------------------------------------------------------------------------------------------------------------------------
+        case storageTypeFile:
+        {
+            // If the backup is compressed then strip off the extension before doing the lookup
+            if (resumeData->compressed)
+                manifestName = strSubN(manifestName, 0, strSize(manifestName) - sizeof(GZIP_EXT));
+
+            // Find the file in both manifests
+            const ManifestFile *file = manifestFileFindDefault(resumeData->manifest, manifestName, NULL);
+            const ManifestFile *fileResume = manifestFileFindDefault(resumeData->manifestResume, manifestName, NULL);
+
+            // Check if the file can be resumed or must be removed
+            const char *removeReason = NULL;
+
+            if (file == NULL)
+                removeReason = "missing in manifest";
+            else if (file->reference != NULL)
+                removeReason = "reference in manifest";
+            else if (fileResume == NULL)
+                removeReason = "missing in resumed manifest";
+            // CSHANG -- this is new and means we don't having to worry about checking for hardlinks on resume because a hardlink
+            // will always have a reference.
+            else if (fileResume->reference != NULL)
+                removeReason = "reference in resumed manifest";
+            else if (fileResume->checksumSha1[0] == '\0')
+                removeReason = "no checksum in resumed manifest";
+            else if (file->size != fileResume->size)
+                removeReason = "mismatched size";
+            else if (!resumeData->delta && file->timestamp != fileResume->timestamp)
+                removeReason = "mismatched timestamp";
+            else if (file->size == 0)
+                // ??? don't resume zero size files because Perl wouldn't -- this can be removed after the migration)
```

Comment:
I don't see where this is true in the Perl code. Can you point me to where this was in the Perl code because as far as I can see in the Perl code was checking for file size equality.

Reply:
In Perl zero-length files did not have checksums in the manifest in memory.  So, checking the reference is defined (https://github.com/pgbackrest/pgbackrest/blob/master/lib/pgBackRest/Backup/Backup.pm#L194) implies the file is also non-zero.

In C we store all checksums in the manifest in memory, and only skip them for zero files when writing to disk.  Since the checksum is always defined we need to check size to mimic the Perl behavior.

Reply:
Ah, ok, I did not realize checksums in C were always defined so I thought the 'removeReason = "no checksum in resumed manifest";' would simply be hit.